File: examples/clients/javascript/icap_client.test.js
const {
    IcapClient,
    IcapConfig,
    HttpRequest,
    HttpResponse,
    IcapResponse,
    IcapError,
    AuthenticationHandler,
    ClientMetrics,
    ICAP_METHODS,
    ICAP_RESPONSE_CODES,
    AUTH_METHODS
} = require('./icap_client');

// Mock axios for testing
jest.mock('axios');
const axios = require('axios');

describe('IcapClient', () => {
    let config;
    let client;

    beforeEach(() => {
        config = new IcapConfig({
            host: '127.0.0.1',
            port: 1344,
            timeout: 1000,
            retries: 0,
            loggingLevel: 'error',
            metricsEnabled: false
        });
        client = new IcapClient(config);
    });

    afterEach(() => {
        if (client) {
            client.close();
        }
    });

    describe('constructor', () => {
        test('should create client with default config', () => {
            const defaultConfig = new IcapConfig();
            const defaultClient = new IcapClient(defaultConfig);
            expect(defaultClient).toBeDefined();
            expect(defaultClient.config.host).toBe('127.0.0.1');
            expect(defaultClient.config.port).toBe(1344);
            defaultClient.close();
        });

        test('should create client with custom config', () => {
            expect(client).toBeDefined();
            expect(client.config.host).toBe('127.0.0.1');
            expect(client.config.port).toBe(1344);
        });

        test('should create client with authentication', () => {
            const authConfig = new IcapConfig({
                authentication: {
                    method: 'basic',
                    username: 'testuser',
                    password: 'testpass'
                }
            });
            const authClient = new IcapClient(authConfig);
            expect(authClient.authHandler).toBeDefined();
            expect(authClient.authHandler.method).toBe('basic');
            authClient.close();
        });
    });

    describe('buildICAPURL', () => {
        test('should build REQMOD URL', () => {
            const url = client.buildICAPURL(ICAP_METHODS.REQMOD);
            expect(url).toBe('icap://127.0.0.1:1344/reqmod');
        });

        test('should build RESPMOD URL', () => {
            const url = client.buildICAPURL(ICAP_METHODS.RESPMOD);
            expect(url).toBe('icap://127.0.0.1:1344/respmod');
        });

        test('should build OPTIONS URL', () => {
            const url = client.buildICAPURL(ICAP_METHODS.OPTIONS);
            expect(url).toBe('icap://127.0.0.1:1344/options');
        });

        test('should throw error for unsupported method', () => {
            expect(() => {
                client.buildICAPURL('INVALID');
            }).toThrow('Unsupported ICAP method: INVALID');
        });
    });

    describe('buildEncapsulatedHeader', () => {
        test('should build header for HTTP request', () => {
            const httpRequest = new HttpRequest();
            const header = client.buildEncapsulatedHeader(httpRequest);
            expect(header).toBe('req-hdr=0, null-body=75');
        });

        test('should build header for HTTP response', () => {
            const httpResponse = new HttpResponse();
            const header = client.buildEncapsulatedHeader(httpResponse);
            expect(header).toBe('res-hdr=0, null-body=120');
        });

        test('should build header for null data', () => {
            const header = client.buildEncapsulatedHeader(null);
            expect(header).toBe('null-body=0');
        });
    });

    describe('serializeHTTPData', () => {
        test('should serialize HTTP request', () => {
            const httpRequest = new HttpRequest({
                method: 'GET',
                uri: '/',
                version: 'HTTP/1.1',
                headers: {
                    'Host': 'example.com',
                    'User-Agent': 'JS-Client'
                },
                body: Buffer.from('test body')
            });

            const data = client.serializeHTTPData(httpRequest);
            const expected = 'GET / HTTP/1.1\r\nHost: example.com\r\nUser-Agent: JS-Client\r\n\r\ntest body';
            expect(data.toString()).toBe(expected);
        });

        test('should serialize HTTP response', () => {
            const httpResponse = new HttpResponse({
                version: 'HTTP/1.1',
                statusCode: 200,
                reason: 'OK',
                headers: {
                    'Content-Type': 'text/html'
                },
                body: Buffer.from('<html>test</html>')
            });

            const data = client.serializeHTTPData(httpResponse);
            const expected = 'HTTP/1.1 200 OK\r\nContent-Type: text/html\r\n\r\n<html>test</html>';
            expect(data.toString()).toBe(expected);
        });
    });

    describe('parseICAPResponse', () => {
        test('should parse ICAP response', () => {
            const responseText = `ICAP/1.0 200 OK
Server: G3ICAP/1.0.0
ISTag: "test-istag"
Methods: REQMOD, RESPMOD, OPTIONS
Service: G3ICAP Content Filter

HTTP/1.1 200 OK
Content-Type: text/html
Content-Length: 13

Hello World!`;

            const response = client.parseICAPResponse(responseText);

            expect(response.version).toBe('ICAP/1.0');
            expect(response.statusCode).toBe(200);
            expect(response.reason).toBe('OK');
            expect(response.headers.Server).toBe('G3ICAP/1.0.0');
            expect(response.headers.ISTag).toBe('"test-istag"');
            expect(response.headers.Methods).toBe('REQMOD, RESPMOD, OPTIONS');
            expect(response.headers.Service).toBe('G3ICAP Content Filter');
        });
    });

    describe('makeRequest', () => {
        beforeEach(() => {
            axios.create.mockReturnValue({
                request: jest.fn()
            });
        });

        test('should make successful request', async () => {
            const mockResponse = {
                status: 200,
                data: `ICAP/1.0 200 OK
Server: G3ICAP/1.0.0

`
            };

            client.httpClient.request.mockResolvedValue(mockResponse);

            const response = await client.makeRequest(ICAP_METHODS.OPTIONS);

            expect(response).toBeDefined();
            expect(response.statusCode).toBe(200);
            expect(client.httpClient.request).toHaveBeenCalledWith({
                method: ICAP_METHODS.OPTIONS,
                url: 'icap://127.0.0.1:1344/options',
                headers: expect.objectContaining({
                    'Host': '127.0.0.1:1344',
                    'User-Agent': 'G3ICAP-JS-Client/1.0.0',
                    'Allow': '204'
                }),
                data: Buffer.alloc(0)
            });
        });

        test('should retry on failure', async () => {
            const mockError = new Error('Network error');
            client.httpClient.request.mockRejectedValue(mockError);

            await expect(client.makeRequest(ICAP_METHODS.OPTIONS)).rejects.toThrow('All retry attempts failed');
            expect(client.httpClient.request).toHaveBeenCalledTimes(1); // No retries in test config
        });
    });

    describe('reqmod', () => {
        test('should send REQMOD request', async () => {
            const mockResponse = {
                status: 200,
                data: `ICAP/1.0 200 OK
Server: G3ICAP/1.0.0

`
            };

            client.httpClient.request.mockResolvedValue(mockResponse);

            const httpRequest = new HttpRequest({
                method: 'GET',
                uri: '/',
                version: 'HTTP/1.1',
                headers: { 'Host': 'example.com' }
            });

            const response = await client.reqmod(httpRequest);

            expect(response).toBeDefined();
            expect(response.statusCode).toBe(200);
        });
    });

    describe('respmod', () => {
        test('should send RESPMOD request', async () => {
            const mockResponse = {
                status: 200,
                data: `ICAP/1.0 200 OK
Server: G3ICAP/1.0.0

`
            };

            client.httpClient.request.mockResolvedValue(mockResponse);

            const httpResponse = new HttpResponse({
                version: 'HTTP/1.1',
                statusCode: 200,
                reason: 'OK',
                headers: { 'Content-Type': 'text/html' }
            });

            const response = await client.respmod(httpResponse);

            expect(response).toBeDefined();
            expect(response.statusCode).toBe(200);
        });
    });

    describe('options', () => {
        test('should send OPTIONS request', async () => {
            const mockResponse = {
                status: 200,
                data: `ICAP/1.0 200 OK
Server: G3ICAP/1.0.0

`
            };

            client.httpClient.request.mockResolvedValue(mockResponse);

            const response = await client.options();

            expect(response).toBeDefined();
            expect(response.statusCode).toBe(200);
        });
    });

    describe('healthCheck', () => {
        test('should return healthy status', async () => {
            const mockResponse = {
                status: 200,
                data: `ICAP/1.0 200 OK
Server: G3ICAP/1.0.0
Methods: REQMOD, RESPMOD, OPTIONS
Service: G3ICAP Content Filter
ISTag: "test-istag"

`
            };

            client.httpClient.request.mockResolvedValue(mockResponse);

            const health = await client.healthCheck();

            expect(health.status).toBe('healthy');
            expect(health.statusCode).toBe(200);
            expect(health.methods).toEqual(['REQMOD', 'RESPMOD', 'OPTIONS']);
            expect(health.version).toBe('G3ICAP Content Filter');
            expect(health.istag).toBe('"test-istag"');
        });

        test('should return unhealthy status on error', async () => {
            const mockError = new Error('Network error');
            client.httpClient.request.mockRejectedValue(mockError);

            const health = await client.healthCheck();

            expect(health.status).toBe('unhealthy');
            expect(health.error).toBe('Network error');
        });
    });
});

describe('AuthenticationHandler', () => {
    test('should handle no authentication', () => {
        const handler = new AuthenticationHandler(AUTH_METHODS.NONE);
        const headers = handler.getHeaders();
        expect(headers).toEqual({});
    });

    test('should handle basic authentication', () => {
        const handler = new AuthenticationHandler(AUTH_METHODS.BASIC, {
            username: 'testuser',
            password: 'testpass'
        });
        const headers = handler.getHeaders();
        expect(headers.Authorization).toBe('Basic dGVzdHVzZXI6dGVzdHBhc3M=');
    });

    test('should handle bearer authentication', () => {
        const handler = new AuthenticationHandler(AUTH_METHODS.BEARER, {
            token: 'test-token'
        });
        const headers = handler.getHeaders();
        expect(headers.Authorization).toBe('Bearer test-token');
    });

    test('should handle JWT authentication', () => {
        const handler = new AuthenticationHandler(AUTH_METHODS.JWT, {
            jwt_token: 'jwt-test-token'
        });
        const headers = handler.getHeaders();
        expect(headers.Authorization).toBe('Bearer jwt-test-token');
    });

    test('should handle API key authentication', () => {
        const handler = new AuthenticationHandler(AUTH_METHODS.API_KEY, {
            api_key: 'test-api-key',
            header_name: 'X-API-Key'
        });
        const headers = handler.getHeaders();
        expect(headers['X-API-Key']).toBe('test-api-key');
    });

    test('should handle API key authentication with default header', () => {
        const handler = new AuthenticationHandler(AUTH_METHODS.API_KEY, {
            api_key: 'test-api-key'
        });
        const headers = handler.getHeaders();
        expect(headers['X-API-Key']).toBe('test-api-key');
    });
});

describe('IcapConfig', () => {
    test('should create config with defaults', () => {
        const config = new IcapConfig();
        expect(config.host).toBe('127.0.0.1');
        expect(config.port).toBe(1344);
        expect(config.timeout).toBe(30000);
        expect(config.retries).toBe(3);
        expect(config.keepAlive).toBe(true);
        expect(config.verifySSL).toBe(true);
        expect(config.loggingLevel).toBe('info');
        expect(config.metricsEnabled).toBe(true);
    });

    test('should create config with custom values', () => {
        const config = new IcapConfig({
            host: 'example.com',
            port: 8080,
            timeout: 60000,
            retries: 5,
            keepAlive: false,
            verifySSL: false,
            loggingLevel: 'debug',
            metricsEnabled: false
        });

        expect(config.host).toBe('example.com');
        expect(config.port).toBe(8080);
        expect(config.timeout).toBe(60000);
        expect(config.retries).toBe(5);
        expect(config.keepAlive).toBe(false);
        expect(config.verifySSL).toBe(false);
        expect(config.loggingLevel).toBe('debug');
        expect(config.metricsEnabled).toBe(false);
    });

    test('should load config from file', () => {
        // Mock fs.readFileSync
        const fs = require('fs');
        const originalReadFileSync = fs.readFileSync;
        fs.readFileSync = jest.fn().mockReturnValue(`
host: example.com
port: 8080
timeout: 60000
`);

        const config = IcapConfig.fromFile('test.yaml');
        expect(config.host).toBe('example.com');
        expect(config.port).toBe(8080);
        expect(config.timeout).toBe(60000);

        // Restore original function
        fs.readFileSync = originalReadFileSync;
    });

    test('should throw error for invalid config file', () => {
        const fs = require('fs');
        const originalReadFileSync = fs.readFileSync;
        fs.readFileSync = jest.fn().mockImplementation(() => {
            throw new Error('File not found');
        });

        expect(() => {
            IcapConfig.fromFile('nonexistent.yaml');
        }).toThrow('Failed to load config from nonexistent.yaml: File not found');

        // Restore original function
        fs.readFileSync = originalReadFileSync;
    });
});

describe('HttpRequest', () => {
    test('should create request with defaults', () => {
        const request = new HttpRequest();
        expect(request.method).toBe('GET');
        expect(request.uri).toBe('/');
        expect(request.version).toBe('HTTP/1.1');
        expect(request.headers).toEqual({});
        expect(request.body).toEqual(Buffer.alloc(0));
    });

    test('should create request with custom values', () => {
        const request = new HttpRequest({
            method: 'POST',
            uri: '/api/test',
            version: 'HTTP/2.0',
            headers: { 'Content-Type': 'application/json' },
            body: Buffer.from('{"test": "data"}')
        });

        expect(request.method).toBe('POST');
        expect(request.uri).toBe('/api/test');
        expect(request.version).toBe('HTTP/2.0');
        expect(request.headers).toEqual({ 'Content-Type': 'application/json' });
        expect(request.body).toEqual(Buffer.from('{"test": "data"}'));
    });
});

describe('HttpResponse', () => {
    test('should create response with defaults', () => {
        const response = new HttpResponse();
        expect(response.version).toBe('HTTP/1.1');
        expect(response.statusCode).toBe(200);
        expect(response.reason).toBe('OK');
        expect(response.headers).toEqual({});
        expect(response.body).toEqual(Buffer.alloc(0));
    });

    test('should create response with custom values', () => {
        const response = new HttpResponse({
            version: 'HTTP/2.0',
            statusCode: 404,
            reason: 'Not Found',
            headers: { 'Content-Type': 'text/html' },
            body: Buffer.from('<html>Not Found</html>')
        });

        expect(response.version).toBe('HTTP/2.0');
        expect(response.statusCode).toBe(404);
        expect(response.reason).toBe('Not Found');
        expect(response.headers).toEqual({ 'Content-Type': 'text/html' });
        expect(response.body).toEqual(Buffer.from('<html>Not Found</html>'));
    });
});

describe('IcapResponse', () => {
    test('should create response with defaults', () => {
        const response = new IcapResponse();
        expect(response.version).toBe('ICAP/1.0');
        expect(response.statusCode).toBe(200);
        expect(response.reason).toBe('OK');
        expect(response.headers).toEqual({});
        expect(response.body).toEqual(Buffer.alloc(0));
        expect(response.httpRequest).toBeNull();
        expect(response.httpResponse).toBeNull();
    });

    test('should create response with custom values', () => {
        const httpRequest = new HttpRequest();
        const httpResponse = new HttpResponse();
        const response = new IcapResponse({
            version: 'ICAP/1.0',
            statusCode: 204,
            reason: 'No Content',
            headers: { 'Server': 'G3ICAP/1.0.0' },
            body: Buffer.from('test'),
            httpRequest,
            httpResponse
        });

        expect(response.version).toBe('ICAP/1.0');
        expect(response.statusCode).toBe(204);
        expect(response.reason).toBe('No Content');
        expect(response.headers).toEqual({ 'Server': 'G3ICAP/1.0.0' });
        expect(response.body).toEqual(Buffer.from('test'));
        expect(response.httpRequest).toBe(httpRequest);
        expect(response.httpResponse).toBe(httpResponse);
    });
});

describe('IcapError', () => {
    test('should create error with message only', () => {
        const error = new IcapError('Test error');
        expect(error.message).toBe('Test error');
        expect(error.name).toBe('IcapError');
        expect(error.code).toBe(0);
        expect(error.cause).toBeNull();
    });

    test('should create error with message and code', () => {
        const error = new IcapError('Test error', 500);
        expect(error.message).toBe('Test error');
        expect(error.code).toBe(500);
        expect(error.cause).toBeNull();
    });

    test('should create error with message, code, and cause', () => {
        const cause = new Error('Original error');
        const error = new IcapError('Test error', 500, cause);
        expect(error.message).toBe('Test error');
        expect(error.code).toBe(500);
        expect(error.cause).toBe(cause);
    });
});

describe('ClientMetrics', () => {
    test('should create metrics', () => {
        const metrics = new ClientMetrics();
        expect(metrics.requestsTotal).toBeDefined();
        expect(metrics.requestsSuccess).toBeDefined();
        expect(metrics.requestsFailed).toBeDefined();
        expect(metrics.responseTime).toBeDefined();
        expect(metrics.connectionPool).toBeDefined();
    });
});

describe('Constants', () => {
    test('should have correct ICAP methods', () => {
        expect(ICAP_METHODS.REQMOD).toBe('REQMOD');
        expect(ICAP_METHODS.RESPMOD).toBe('RESPMOD');
        expect(ICAP_METHODS.OPTIONS).toBe('OPTIONS');
    });

    test('should have correct ICAP response codes', () => {
        expect(ICAP_RESPONSE_CODES.CONTINUE).toBe(100);
        expect(ICAP_RESPONSE_CODES.OK).toBe(200);
        expect(ICAP_RESPONSE_CODES.NO_CONTENT).toBe(204);
        expect(ICAP_RESPONSE_CODES.BAD_REQUEST).toBe(400);
        expect(ICAP_RESPONSE_CODES.NOT_FOUND).toBe(404);
        expect(ICAP_RESPONSE_CODES.INTERNAL_SERVER_ERROR).toBe(500);
    });

    test('should have correct authentication methods', () => {
        expect(AUTH_METHODS.NONE).toBe('none');
        expect(AUTH_METHODS.BASIC).toBe('basic');
        expect(AUTH_METHODS.BEARER).toBe('bearer');
        expect(AUTH_METHODS.JWT).toBe('jwt');
        expect(AUTH_METHODS.API_KEY).toBe('api_key');
    });
});

File: examples/clients/go/icap_client_test.go
package main

import (
	"context"
	"testing"
	"time"
)

// TestIcapClient_NewIcapClient tests client creation
func TestIcapClient_NewIcapClient(t *testing.T) {
	config := &IcapConfig{
		Host:               "127.0.0.1",
		Port:               1344,
		Timeout:            30 * time.Second,
		Retries:            3,
		RetryDelay:         time.Second,
		MaxRetryDelay:      60 * time.Second,
		BackoffFactor:      2.0,
		ConnectionPoolSize: 10,
		KeepAlive:          true,
		VerifySSL:          true,
		LoggingLevel:       "INFO",
		MetricsEnabled:     true,
	}

	client := NewIcapClient(config)
	if client == nil {
		t.Fatal("Expected client to be created")
	}

	if client.config.Host != config.Host {
		t.Errorf("Expected host %s, got %s", config.Host, client.config.Host)
	}

	if client.config.Port != config.Port {
		t.Errorf("Expected port %d, got %d", config.Port, client.config.Port)
	}

	client.Close()
}

// TestIcapClient_NewIcapClientWithAuth tests client creation with authentication
func TestIcapClient_NewIcapClientWithAuth(t *testing.T) {
	config := &IcapConfig{
		Host:               "127.0.0.1",
		Port:               1344,
		Timeout:            30 * time.Second,
		Retries:            3,
		RetryDelay:         time.Second,
		MaxRetryDelay:      60 * time.Second,
		BackoffFactor:      2.0,
		ConnectionPoolSize: 10,
		KeepAlive:          true,
		VerifySSL:          true,
		LoggingLevel:       "INFO",
		MetricsEnabled:     true,
		Authentication: map[string]string{
			"method":   "basic",
			"username": "testuser",
			"password": "testpass",
		},
	}

	client := NewIcapClient(config)
	if client == nil {
		t.Fatal("Expected client to be created")
	}

	if client.authHandler == nil {
		t.Fatal("Expected auth handler to be created")
	}

	if client.authHandler.method != AuthBasic {
		t.Errorf("Expected auth method %s, got %s", AuthBasic, client.authHandler.method)
	}

	client.Close()
}

// TestAuthenticationHandler_GetHeaders tests authentication header generation
func TestAuthenticationHandler_GetHeaders(t *testing.T) {
	tests := []struct {
		name     string
		method   AuthenticationMethod
		config   map[string]string
		expected map[string]string
	}{
		{
			name:   "No authentication",
			method: AuthNone,
			config: map[string]string{},
			expected: map[string]string{},
		},
		{
			name:   "Basic authentication",
			method: AuthBasic,
			config: map[string]string{
				"username": "testuser",
				"password": "testpass",
			},
			expected: map[string]string{
				"Authorization": "Basic dGVzdHVzZXI6dGVzdHBhc3M=",
			},
		},
		{
			name:   "Bearer authentication",
			method: AuthBearer,
			config: map[string]string{
				"token": "test-token",
			},
			expected: map[string]string{
				"Authorization": "Bearer test-token",
			},
		},
		{
			name:   "JWT authentication",
			method: AuthJWT,
			config: map[string]string{
				"jwt_token": "jwt-test-token",
			},
			expected: map[string]string{
				"Authorization": "Bearer jwt-test-token",
			},
		},
		{
			name:   "API Key authentication",
			method: AuthAPIKey,
			config: map[string]string{
				"api_key":    "test-api-key",
				"header_name": "X-API-Key",
			},
			expected: map[string]string{
				"X-API-Key": "test-api-key",
			},
		},
		{
			name:   "API Key authentication with default header",
			method: AuthAPIKey,
			config: map[string]string{
				"api_key": "test-api-key",
			},
			expected: map[string]string{
				"X-API-Key": "test-api-key",
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			handler := NewAuthenticationHandler(tt.method, tt.config)
			headers := handler.GetHeaders()

			if len(headers) != len(tt.expected) {
				t.Errorf("Expected %d headers, got %d", len(tt.expected), len(headers))
			}

			for name, expectedValue := range tt.expected {
				if actualValue, ok := headers[name]; !ok {
					t.Errorf("Expected header %s not found", name)
				} else if actualValue != expectedValue {
					t.Errorf("Expected header %s value %s, got %s", name, expectedValue, actualValue)
				}
			}
		})
	}
}

// TestIcapClient_buildICAPURL tests URL building
func TestIcapClient_buildICAPURL(t *testing.T) {
	config := &IcapConfig{
		Host: "127.0.0.1",
		Port: 1344,
	}

	client := NewIcapClient(config)
	defer client.Close()

	tests := []struct {
		method   IcapMethod
		expected string
	}{
		{REQMOD, "icap://127.0.0.1:1344/reqmod"},
		{RESPMOD, "icap://127.0.0.1:1344/respmod"},
		{OPTIONS, "icap://127.0.0.1:1344/options"},
	}

	for _, tt := range tests {
		t.Run(string(tt.method), func(t *testing.T) {
			url := client.buildICAPURL(tt.method)
			if url != tt.expected {
				t.Errorf("Expected URL %s, got %s", tt.expected, url)
			}
		})
	}
}

// TestIcapClient_buildEncapsulatedHeader tests encapsulated header building
func TestIcapClient_buildEncapsulatedHeader(t *testing.T) {
	config := &IcapConfig{}
	client := NewIcapClient(config)
	defer client.Close()

	// Test with HTTP request
	httpRequest := &HttpRequest{
		Method:  "GET",
		URI:     "/",
		Version: "HTTP/1.1",
		Headers: map[string]string{
			"Host": "example.com",
		},
	}

	header := client.buildEncapsulatedHeader(httpRequest)
	if header != "req-hdr=0, null-body=75" {
		t.Errorf("Expected 'req-hdr=0, null-body=75', got %s", header)
	}

	// Test with HTTP response
	httpResponse := &HttpResponse{
		Version:    "HTTP/1.1",
		StatusCode: 200,
		Reason:     "OK",
		Headers: map[string]string{
			"Content-Type": "text/html",
		},
	}

	header = client.buildEncapsulatedHeader(httpResponse)
	if header != "res-hdr=0, null-body=120" {
		t.Errorf("Expected 'res-hdr=0, null-body=120', got %s", header)
	}

	// Test with nil
	header = client.buildEncapsulatedHeader(nil)
	if header != "null-body=0" {
		t.Errorf("Expected 'null-body=0', got %s", header)
	}
}

// TestIcapClient_serializeHTTPData tests HTTP data serialization
func TestIcapClient_serializeHTTPData(t *testing.T) {
	config := &IcapConfig{}
	client := NewIcapClient(config)
	defer client.Close()

	// Test HTTP request serialization
	httpRequest := &HttpRequest{
		Method:  "GET",
		URI:     "/",
		Version: "HTTP/1.1",
		Headers: map[string]string{
			"Host":       "example.com",
			"User-Agent": "Go-Client",
		},
		Body: []byte("test body"),
	}

	data := client.serializeHTTPData(httpRequest)
	expected := "GET / HTTP/1.1\r\nHost: example.com\r\nUser-Agent: Go-Client\r\n\r\ntest body"
	if string(data) != expected {
		t.Errorf("Expected %s, got %s", expected, string(data))
	}

	// Test HTTP response serialization
	httpResponse := &HttpResponse{
		Version:    "HTTP/1.1",
		StatusCode: 200,
		Reason:     "OK",
		Headers: map[string]string{
			"Content-Type": "text/html",
		},
		Body: []byte("<html>test</html>"),
	}

	data = client.serializeHTTPData(httpResponse)
	expected = "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\n\r\n<html>test</html>"
	if string(data) != expected {
		t.Errorf("Expected %s, got %s", expected, string(data))
	}
}

// TestIcapClient_parseICAPResponse tests ICAP response parsing
func TestIcapClient_parseICAPResponse(t *testing.T) {
	config := &IcapConfig{}
	client := NewIcapClient(config)
	defer client.Close()

	responseText := `ICAP/1.0 200 OK
Server: G3ICAP/1.0.0
ISTag: "test-istag"
Methods: REQMOD, RESPMOD, OPTIONS
Service: G3ICAP Content Filter

HTTP/1.1 200 OK
Content-Type: text/html
Content-Length: 13

Hello World!`

	response := client.parseICAPResponse(responseText)

	if response.Version != "ICAP/1.0" {
		t.Errorf("Expected version ICAP/1.0, got %s", response.Version)
	}

	if response.StatusCode != 200 {
		t.Errorf("Expected status code 200, got %d", response.StatusCode)
	}

	if response.Reason != "OK" {
		t.Errorf("Expected reason OK, got %s", response.Reason)
	}

	expectedHeaders := map[string]string{
		"Server":   "G3ICAP/1.0.0",
		"ISTag":    "\"test-istag\"",
		"Methods":  "REQMOD, RESPMOD, OPTIONS",
		"Service":  "G3ICAP Content Filter",
	}

	for name, expectedValue := range expectedHeaders {
		if actualValue, ok := response.Headers[name]; !ok {
			t.Errorf("Expected header %s not found", name)
		} else if actualValue != expectedValue {
			t.Errorf("Expected header %s value %s, got %s", name, expectedValue, actualValue)
		}
	}

	expectedBody := "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\nContent-Length: 13\r\n\r\nHello World!"
	if string(response.Body) != expectedBody {
		t.Errorf("Expected body %s, got %s", expectedBody, string(response.Body))
	}
}

// TestIcapClient_Options tests OPTIONS method
func TestIcapClient_Options(t *testing.T) {
	// This test would require a running ICAP server
	// For now, we'll just test that the method exists and doesn't panic
	config := &IcapConfig{
		Host:           "127.0.0.1",
		Port:           1344,
		Timeout:        time.Second,
		Retries:        0, // No retries for test
		LoggingLevel:   "ERROR", // Reduce logging noise
		MetricsEnabled: false,
	}

	client := NewIcapClient(config)
	defer client.Close()

	ctx := context.Background()
	_, err := client.Options(ctx)
	
	// We expect this to fail since there's no server running
	// but we're just testing that the method doesn't panic
	if err == nil {
		t.Log("OPTIONS request succeeded (unexpected in test environment)")
	} else {
		t.Logf("OPTIONS request failed as expected: %v", err)
	}
}

// TestIcapClient_Reqmod tests REQMOD method
func TestIcapClient_Reqmod(t *testing.T) {
	config := &IcapConfig{
		Host:           "127.0.0.1",
		Port:           1344,
		Timeout:        time.Second,
		Retries:        0, // No retries for test
		LoggingLevel:   "ERROR", // Reduce logging noise
		MetricsEnabled: false,
	}

	client := NewIcapClient(config)
	defer client.Close()

	httpRequest := &HttpRequest{
		Method:  "GET",
		URI:     "/",
		Version: "HTTP/1.1",
		Headers: map[string]string{
			"Host": "example.com",
		},
	}

	ctx := context.Background()
	_, err := client.Reqmod(ctx, httpRequest)
	
	// We expect this to fail since there's no server running
	// but we're just testing that the method doesn't panic
	if err == nil {
		t.Log("REQMOD request succeeded (unexpected in test environment)")
	} else {
		t.Logf("REQMOD request failed as expected: %v", err)
	}
}

// TestIcapClient_Respmod tests RESPMOD method
func TestIcapClient_Respmod(t *testing.T) {
	config := &IcapConfig{
		Host:           "127.0.0.1",
		Port:           1344,
		Timeout:        time.Second,
		Retries:        0, // No retries for test
		LoggingLevel:   "ERROR", // Reduce logging noise
		MetricsEnabled: false,
	}

	client := NewIcapClient(config)
	defer client.Close()

	httpResponse := &HttpResponse{
		Version:    "HTTP/1.1",
		StatusCode: 200,
		Reason:     "OK",
		Headers: map[string]string{
			"Content-Type": "text/html",
		},
		Body: []byte("<html>test</html>"),
	}

	ctx := context.Background()
	_, err := client.Respmod(ctx, httpResponse)
	
	// We expect this to fail since there's no server running
	// but we're just testing that the method doesn't panic
	if err == nil {
		t.Log("RESPMOD request succeeded (unexpected in test environment)")
	} else {
		t.Logf("RESPMOD request failed as expected: %v", err)
	}
}

// TestIcapClient_HealthCheck tests health check
func TestIcapClient_HealthCheck(t *testing.T) {
	config := &IcapConfig{
		Host:           "127.0.0.1",
		Port:           1344,
		Timeout:        time.Second,
		Retries:        0, // No retries for test
		LoggingLevel:   "ERROR", // Reduce logging noise
		MetricsEnabled: false,
	}

	client := NewIcapClient(config)
	defer client.Close()

	ctx := context.Background()
	health, err := client.HealthCheck(ctx)
	
	// We expect this to fail since there's no server running
	// but we're just testing that the method doesn't panic
	if err != nil {
		t.Logf("Health check failed as expected: %v", err)
		return
	}

	if health["status"] != "unhealthy" {
		t.Errorf("Expected status unhealthy, got %s", health["status"])
	}
}

// TestLoadConfig tests configuration loading
func TestLoadConfig(t *testing.T) {
	// Test with invalid config file
	_, err := LoadConfig("nonexistent.yaml")
	if err == nil {
		t.Error("Expected error for nonexistent config file")
	}

	// Test with valid config (would need to create a test config file)
	// For now, we'll just test that the function exists
	if LoadConfig == nil {
		t.Error("LoadConfig function not found")
	}
}

// TestIcapError tests error handling
func TestIcapError(t *testing.T) {
	err := &IcapError{
		Message: "Test error",
		Code:    500,
		Err:     nil,
	}

	if err.Error() != "Test error" {
		t.Errorf("Expected error message 'Test error', got %s", err.Error())
	}

	errWithCause := &IcapError{
		Message: "Test error",
		Code:    500,
		Err:     err,
	}

	expected := "Test error: Test error"
	if errWithCause.Error() != expected {
		t.Errorf("Expected error message '%s', got %s", expected, errWithCause.Error())
	}
}

// TestClientMetrics tests metrics creation
func TestClientMetrics(t *testing.T) {
	metrics := NewClientMetrics()
	if metrics == nil {
		t.Fatal("Expected metrics to be created")
	}

	if metrics.RequestsTotal == nil {
		t.Error("Expected RequestsTotal counter to be created")
	}

	if metrics.RequestsSuccess == nil {
		t.Error("Expected RequestsSuccess counter to be created")
	}

	if metrics.RequestsFailed == nil {
		t.Error("Expected RequestsFailed counter to be created")
	}

	if metrics.ResponseTime == nil {
		t.Error("Expected ResponseTime histogram to be created")
	}

	if metrics.ConnectionPool == nil {
		t.Error("Expected ConnectionPool gauge to be created")
	}
}

// BenchmarkIcapClient_serializeHTTPData benchmarks HTTP data serialization
func BenchmarkIcapClient_serializeHTTPData(b *testing.B) {
	config := &IcapConfig{}
	client := NewIcapClient(config)
	defer client.Close()

	httpRequest := &HttpRequest{
		Method:  "GET",
		URI:     "/",
		Version: "HTTP/1.1",
		Headers: map[string]string{
			"Host":       "example.com",
			"User-Agent": "Go-Client",
		},
		Body: []byte("test body"),
	}

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		client.serializeHTTPData(httpRequest)
	}
}

// BenchmarkIcapClient_parseICAPResponse benchmarks ICAP response parsing
func BenchmarkIcapClient_parseICAPResponse(b *testing.B) {
	config := &IcapConfig{}
	client := NewIcapClient(config)
	defer client.Close()

	responseText := `ICAP/1.0 200 OK
Server: G3ICAP/1.0.0
ISTag: "test-istag"
Methods: REQMOD, RESPMOD, OPTIONS
Service: G3ICAP Content Filter

HTTP/1.1 200 OK
Content-Type: text/html
Content-Length: 13

Hello World!`

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		client.parseICAPResponse(responseText)
	}
}
